Feature/package references (#71)
madewithkoji/docs: src/asciidoc/reference/packages/withkojiDatabase.adoc

implement a Koji database for the backend of your template.
//end::description[]

Koji databases are made up of one or more `collections`.
Each `collection` stores a series of database entries with a unique key and collection of key-value pairs.

> Values within the key-value pairs can be of any type, including strings, numbers and arrays.

Reviewer: ```suggestion
Values within the key-value pairs can be of any type, including strings, numbers, and arrays.
```

implement a Koji database for the backend of your template.
//end::description[]

> 

Author: See https://github.com/madewithkoji/docs/pull/71#discussion_r494673014

> 

Reviewer: Same thing here... A sentence or two about the structure, since it is prescribed, would be good to know. That info would allow a developer to understand how these methods work and evaluate whether this is the right package for them.

[.hcode, id=".arrayRemove", reftext="arrayRemove"]
=== .arrayRemove(collection, key, value)

Removes data from arrays in an existing database entry.

==== Parameters

* `collection` – _String_, name of the collection
* `key` – _String_, key for the database entry.
* `value` – _Any_, key-value pairs of arrays and the entries to remove from them.

==== Returns

(Async) _Boolean_, indicates whether the data was successfully removed from arrays.

==== Example

[source, javascript]
const isRemoved = await database.arrayRemove('myCollection', 'myExistingKey', {'myValue':1});

[.hcode, id=".delete", reftext="delete"]
=== .delete(collection, key)

Deletes an entire entry from the database.

==== Parameters

* `collection` – _String_, name of the collection.
* `key` – _String_, key of the database entry.

==== Returns

(Async) _Boolean_, indicates whether the entry was successfully deleted.

==== Example

[source, javascript]
const isDeleted = await database.delete('myCollection', 'myExistingKey');

[.hcode, id=".generateSignedUploadRequest", reftext="generateSignedUploadRequest"]
=== .generateSignedUploadRequest(fileName)

> Generates a signed URL for uploading a file directly to the your project’s CDN.

Author: Waiting to hear back from Sean/diddy on this one. Not really sure the use-case for this function (also currently returns error)

=== .getCollections()

Gets all the collections available in the database.

==== Returns

(Async) Array of _Strings_, list containing the names of the collections.

==== Example

[source, javascript]
const collections = await database.getCollections();

[.hcode, id=".getWhere", reftext="getWhere"]
=== .getWhere(collection, predicateKey, predicateOperation, predicateValue)

Gets all the database entries that match a query against a value.

==== Parameters

* `collection` – _String_, name of the collection.
* `predicateKey` – _String_, key of the data to query against.
* `predicateOperation` – _String_, operation to query the data against, possible values are `<`, `\<=`, `==`, `>`, `>=`.
* `predicateValue` – _String_, value to run the query against.

==== Returns

(Async) Array of _Objects_, list of the database entries that match the query.

==== Example

[source, javascript]
const results = await database.getWhere('myCollection','predicateKey', '==', 'predicateValue');

[.hcode, id=".search", reftext="search"]
=== .search(collection, queryKey, queryValue)

Gets all the database entries that have a partial match on a particular key

> NOTE: The `queryKey` must be set for all entries in your data, if some entries have no value for that key it cannot be searched on.

Author: Perhaps wrote this in a misleading way. It doesn't return an error, it just doesn't return anything.
This may be a bug in the package (or the underlying database service) as it seems like odd functionality.

==== Parameters

* `collection` – _String_, name of the collection.
* `queryKey` – _String_, key to partially match against.
* `queryValue` – _String_, value for the partial match.

==== Returns

(Async) Array of _Objects_, list of database entries that have a partial match.

==== Example

[source, javascript]
const results = await database.search('myCollection','myKey', 'myValue');

[.hcode, id=".set", reftext="set"]
=== .set(collection, key, value)

Adds an entry to the database.

==== Parameters

* `collection` – _String_, name of the collection.

> * `key` – _String_, key for the value.

Reviewer: ```suggestion
* `key` – _String_, key of the database entry.
```

> * `value` – _Any_, key-value pairs to add to the database.

Reviewer: ```suggestion
* `data` – _Any_, key-value pairs to add to the database.
```

==== Returns

(Async) _Boolean_, indicates whether the entry was successfully added.

==== Example

[source,javascript]
----
const isAdded = await database.set('myCollection', 'myKey', {'myValue':1});
----

[.hcode, id=".update", reftext="update"]
=== .update(collection, key, value)

Updates an entry in the database with the given value.

==== Parameters

* `collection` – _String_, name of the collection.
* `key` – _String_, key for the value in the database

> * `value` – _Any_, key-value pairs to update on the value.

Author: No, only updates specified keys. Will add to the description.

NOTE: The `searchAttribute` must have a value set for all entries in your database.
If some entries have no value for that key, nothing will be returned.

==== Parameters

* `collection` – _String_, name of the collection.
* `searchAttribute` – _String_, key to partially match against.
* `searchValue` – _String_, value for the partial match.

==== Returns

(Async) Array of _Objects_, list of database entries that have a partial match.

==== Example

[source, javascript]
const results = await database.search('myCollection','myKey', 'myValue');

[.hcode, id=".set", reftext="set"]
=== .set(collection, key, value)

Adds an entry to the database.

==== Parameters

* `collection` – _String_, name of the collection.
* `key` – _String_, key for the value.
* `value` – _Any_, key-value pairs to add to the database.

==== Returns

(Async) _Boolean_, indicates whether the entry was successfully added.

==== Example

[source,javascript]
----
const isAdded = await database.set('myCollection', 'myKey', {'myValue':1});
----

[.hcode, id=".update", reftext="update"]
=== .update(collection, key, data)

Updates an entry in the database with the given value.

> NOTE: Only updates specified values, anything not present in `data` will be left unchanged.

Reviewer: ```suggestion
NOTE: This method updates only the values specified in `data`. If additional values exist for the key, they are not changed.
```
